### PR TITLE
[Enhance] Avoid unnecessary `listdir` when building `ImageNet`

### PR DESCRIPTION
**Pull request by sovrasov**
## Motivation
Implementation of the `ImageNet` dataset contains duplicated operations: one extra call of `os.listdir` and calls of `os.path.isdir` for every class folder in the dataset. Since this loader can be used for much more larger datasets than ImageNet, it should be as effective as possible. In pytorch's ImageFolder dataset which is basically pretty close to the considered `ImageNet` class, these duplications are eliminated. 

## Modification

Duplicated calls were eliminated (all the top level folder structure checks are concentrated in `find_folders`).

## BC-breaking (Optional)

This PR doesn't brake anything.

### CODE REVIEW

**Comment by mzr1996** [comment]
Thanks for your contribution! We will review it recently.

File: mmcls/datasets/imagenet.py [hunk 1]
```diff
@@ -49,11 +49,8 @@ def get_samples(root, folder_to_idx, extensions):
     """
     samples = []
     root = os.path.expanduser(root)
-    for folder_name in sorted(os.listdir(root)):
+    for folder_name in sorted(list(folder_to_idx.keys())):
```
Comment (mzr1996) [inline mmcls/datasets/imagenet.py:52]: After the change, we need to assume all folders in `folder_to_idx` exist.
Considering it's an inner function, and only be used after `find_folders`, I think it's a reasonable assumption.

**Review by mzr1996 — APPROVED** [review]
Looks good.

**Review by Ezra-Yu — APPROVED** [review]
LGTM

**Pull request merged**